lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -98,13 +111,130 @@ import qualified Data.Map.Merge.Strict as Map
 import qualified Data.Map.Strict as Map
 import qualified Data.Text as T
 
--- | Stake Pool Data fields fetched from the node via LSQ
+--
+-- Stake Pool Layer
+--
+
+data StakePoolLayer = StakePoolLayer
+    { knownPools
+        :: IO [PoolId]
+
+    -- | List pools based given the the amount of stake the user intends to
+    -- delegate, which affects the size of the rewards and the ranking of the
+    -- pools.
+    , listStakePools
+        :: Coin
+        -> ExceptT ErrNetworkUnavailable IO [Api.ApiStakePool]
+    }
+
+newStakePoolLayer
+    :: GenesisParameters
+    -> NetworkLayer IO (IO Shelley) b
+    -> DBLayer IO
+    -> StakePoolLayer
+newStakePoolLayer gp nl db = StakePoolLayer
+    { knownPools = _knownPools
+    , listStakePools = _listPools
+    }
+  where
+    _knownPools
+        :: IO [PoolId]
+    _knownPools = do
+        tip <- getTip
+        let dummyCoin = Coin 0
+        res <- runExceptT $ map fst . Map.toList
+            . combineLsqData <$> stakeDistribution nl tip dummyCoin
```

Comment:
known pools should be implemented using the _database_ only. That is its sole purpose, otherwise it is utterly redundant with listPools. The goal is to have a fast access to known pools ids to avoid having to query the network layer when delegating / undelegating to pools.

Reply:
What do you say about: https://github.com/input-output-hk/cardano-wallet/pull/1753#discussion_r440780076 ?

> I know knownPools are used to validate the users's choice of delegation.
>
> The pool worker is less likely to be in sync with the network (more downstream) than the node.
>
> Wouldn't implementing it in terms of the LSQ make more sense?

Though perhaps, having the pool worker out of sync with the node is unlikely, since they Daedalus users will always launch the node and wallet at the same time?

Reply:
Because that is why we have this function in the first place. We can already get the full list of pool ids from the `listPools` and it was how it was done _before_. Yet, this was slow and sometimes _unreliable_ because it needed to go through the much more work. As a result, requests for joining / quitting pools were often failing because we couldn't reliably get a list of pools ids. 

Yet, getting the list of pool ids is straightforward and very reliable when fetched from the local database. Implementing this function using the LSQ completely defies its original purpose which was to circumvent the network layer!

Reply:
In jormungandr we required the DB to be in sync with the node. For haskell we don't.

If the user ends up with a synced node and no stake-pools.sqlite, and lists pools, he will see pools based on LSQ data from the node tip. If he were to try to delegate, I fear the wallet wouldn't let him, claiming that the pool he selected doesn't exist.

It just seems weird to me that we would treat the node as the source of [truth](https://github.com/input-output-hk/cardano-wallet/blob/2f1a21f67f1ab3f5ae8e57da5c97842bf0d0532d/lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs#L201), but then validate the delegation choice based a different one.

Reply:
> If he were to try to delegate, I fear the wallet wouldn't let him, claiming that the pool he selected doesn't exist.

Which is fine, the pool worker won't be _that_ behind and will catch up quite quickly. We can leave it as it is; if the networking layer is fast enough and reliable, then it _should be okay_. Let's see.